hudi-utilities/src/main/java/org/apache/hudi/utilities/keygen/TimestampBasedComplexKeyGenerator.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.utilities.keygen;
+
+import org.apache.hudi.DataSourceUtils;
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.exception.HoodieKeyException;
+import org.apache.hudi.keygen.ComplexKeyGenerator;
+import org.apache.hudi.utilities.exception.HoodieDeltaStreamerException;
+import org.apache.hudi.utilities.keygen.common.KeyGeneratorTimestampConfig;
+import org.apache.hudi.utilities.keygen.common.TimestampTypeEnum;
+import org.apache.hudi.utilities.keygen.util.KeyGeneratorTimestampUtil;
+
+import org.apache.avro.generic.GenericRecord;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Date;
+import java.util.TimeZone;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Key generator, that relies on timestamps for partitioning field. Still picks record key by name.
+ */
+public class TimestampBasedComplexKeyGenerator extends ComplexKeyGenerator {
```

Comment:
IMO, the `TimestampBasedKeyGenerator` and `TimestampBasedComplexKeyGenerator` have the similar business logic(both in constructor and timestamp parse). Can we extract them with an abstract class so that we can reduce some unnecessary code?

Reply:
@yanghua that's a good suggestion, but a little diffcult. because `TimestampBasedComplexKeyGenerator `has already extends `ComplexKeygenerator`, I'll see what I can do to reuse more code.
thanks for your review